dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/process/ProcessServiceImpl.java

```diff
@@ -462,6 +462,40 @@ public ProcessDefinition findProcessDefinition(Long processDefinitionCode, int v
         return processDefinition;
     }
 
+    /**
+     * find a batch of process definitions by a map of <code, version>.
+     * @param codeVersionMap Map<code, version>
+     * @return
+     */
+    @Override
+    public List<ProcessDefinition> findProcessDefinitions(Map<Long, Integer> codeVersionMap) {
+        Set<Long> codes = codeVersionMap.keySet();
+        List<ProcessDefinition> processDefinitions = processDefineMapper.queryByCodes(codes);
+        Map<Long, Integer> codeVersionNeedToRetrieve = new HashMap<>();
+        codeVersionNeedToRetrieve.putAll(codeVersionMap);
+        // filter out the code/version entry that don't need to retrieve
+        processDefinitions.forEach(processDefinition -> {
+            long code = processDefinition.getCode();
+            if (codeVersionNeedToRetrieve.containsKey(code) && processDefinition.getVersion() == codeVersionNeedToRetrieve.get(code).intValue()) {
+                codeVersionNeedToRetrieve.remove(code);
+            }
+        });
+        if (!codeVersionNeedToRetrieve.isEmpty()) {
+            List<ProcessDefinition> complementProcessDefinitions = codeVersionNeedToRetrieve.entrySet()
+                    .stream()
+                    .map(entry -> {
+                        ProcessDefinition processDefinition = processDefineLogMapper.queryByDefinitionCodeAndVersion(entry.getKey(), entry.getValue());
+                        if (processDefinition != null) {
+                            processDefinition.setId(0);
+                        }
+                        return processDefinition;
+                    })
+                    .collect(Collectors.toList());
+            processDefinitions.addAll(complementProcessDefinitions);
+        }
+        return processDefinitions;
```

Comment:
Can we directly query from `processDefineLogMapper` by using below SQL?
```
select * from t_ds_process_definition_log where (code = xx and version = xx) or (code = xxx and version = xxx)
```

Reply:
Yeah, got it. It's a better idea.